File: python/prism_sdk/autonomous_goal_recovery.py
"""Ordered, metadata-only recovery for long-horizon autonomous goals.

The worker journal and control-loop checkpoint are useful independently, but restoring them in
the wrong order can admit a new cycle while an older provider/effect invocation is uncertain.
This module composes the two caller-owned persistence coordinators: journal restoration happens
first, active boundaries are reconciled and flushed, and only then is the loop checkpoint exposed
for resume. Task text, prompts, parameters, credentials, provider values, evaluator payloads,
and results never enter the report.
"""

from __future__ import annotations

from collections.abc import Mapping
import copy
import re
from typing import Any, Literal

from .authoring import canonical_json, content_digest
from .autonomous_goal_control_loop import AutonomousGoalControlLoop, AutonomousGoalControlLoopResult
from .autonomous_goal_control_persistence import (
    AutonomousGoalControlLoopPersistenceCoordinator,
    validate_autonomous_goal_control_loop_snapshot,
)
from .autonomous_goal_worker_journal import (
    AutonomousGoalWorkerJournalPersistenceCoordinator,
)
from .goals import AutonomousGoalError, AutonomousGoalLedger


GOAL_RECOVERY_SCHEMA = "bioprism-autonomous-goal-recovery/0.1"
GOAL_RECOVERY_RETENTION = "metadata_only_goal_recovery;tasks_prompts_parameters_credentials_provider_values_and_results_not_retained"
MAX_GOAL_RECOVERY_GOALS = 16_384
MAX_GOAL_RECOVERY_REPORT_BYTES = 2_000_000
_DIGEST = re.compile(r"^[0-9a-f]{64}$")

RecoveryStatus = Literal["fresh", "restored", "recovered"]


def _fail(message: str) -> None:
    raise AutonomousGoalError(f"autonomous goal recovery {message}")


def _digest(value: Any, *, name: str, allow_none: bool = False) -> str | None:
    if value is None and allow_none:
        return None
    if not isinstance(value, str) or _DIGEST.fullmatch(value) is None:
        _fail(f"{name} must be a lowercase SHA-256 digest")
    return value


def _integer(value: Any, *, name: str, minimum: int = 0, maximum: int = 2**63 - 1) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value < minimum or value > maximum:
        _fail(f"{name} is outside its integer bounds")
    return value


def _identifier(value: Any, *, name: str, maximum: int = 256) -> str:
    if not isinstance(value, str) or not value.strip() or "\x00" in value or len(value.encode("utf-8")) > maximum:
        _fail(f"{name} is outside its bounded identifier contract")
    return value.strip()


def _entry(value: Any, *, index: int) -> dict[str, Any]:
    if not isinstance(value, Mapping):
        _fail(f"recovered entry {index} is malformed")
    expected = {"goal_id", "from_phase", "goal_status", "outcome_digest"}
    if set(value) != expected:
        _fail(f"recovered entry {index} contains unsupported or missing fields")
    if value["from_phase"] not in {"claimed", "dispatch_started"}:
        _fail(f"recovered entry {index} phase is invalid")
    if value["goal_status"] not in {"paused", "blocked"}:
        _fail(f"recovered entry {index} status is invalid")
    return {
        "goal_id": _identifier(value["goal_id"], name=f"recovered entry {index} goal_id"),
        "from_phase": value["from_phase"],
        "goal_status": value["goal_status"],
        "outcome_digest": _digest(value["outcome_digest"], name=f"recovered entry {index} outcome_digest"),
    }


def _report_body(value: Any, *, require_digest: bool) -> dict[str, Any]:
    if not isinstance(value, Mapping):
        _fail("report must be a mapping")
    required = {
        "schema", "status", "active_count_before_recovery", "recovered", "recovery_digest", "journal_snapshot_digest",
        "journal_head_digest", "control_loop_snapshot_digest", "control_loop_generation", "resume_snapshot", "ready_to_resume",
        "requires_external_reconciliation", "retention", "secret_material",
    }
    allowed = required | {"report_digest"}
    if set(value).difference(allowed) or not required.issubset(value) or (require_digest and "report_digest" not in value):
        _fail("report contains unsupported or missing fields")
    if value["schema"] != GOAL_RECOVERY_SCHEMA or value["retention"] != GOAL_RECOVERY_RETENTION or value["secret_material"] != "never_returned":
        _fail("report markers are invalid")
    if value["status"] not in {"fresh", "restored", "recovered"}:
        _fail("report status is invalid")
    active_count = _integer(value["active_count_before_recovery"], name="active_count_before_recovery", maximum=MAX_GOAL_RECOVERY_GOALS)
    raw_recovered = value["recovered"]
    if not isinstance(raw_recovered, list) or len(raw_recovered) > MAX_GOAL_RECOVERY_GOALS:
        _fail("recovered entries are outside their bounds")
    recovered = [_entry(raw, index=index) for index, raw in enumerate(raw_recovered)]
    if len(recovered) != active_count:
        _fail("recovered entries do not account for every active journal boundary")
    goal_ids = [row["goal_id"] for row in recovered]
    if len(set(goal_ids)) != len(goal_ids):
        _fail("recovered entries contain duplicate goals")
    recovery_digest = _digest(value["recovery_digest"], name="recovery_digest")
    if recovery_digest != content_digest(recovered):
        _fail("recovery digest does not match recovered entries")
    journal_snapshot_digest = _digest(value["journal_snapshot_digest"], name="journal_snapshot_digest", allow_none=True)
    journal_head_digest = value["journal_head_digest"] if value["journal_head_digest"] == "" else _digest(value["journal_head_digest"], name="journal_head_digest")
    control_snapshot_digest = _digest(value["control_loop_snapshot_digest"], name="control_loop_snapshot_digest", allow_none=True)
    control_generation = _integer(value["control_loop_generation"], name="control_loop_generation", maximum=2**31 - 1)
    resume_snapshot = None if value["resume_snapshot"] is None else validate_autonomous_goal_control_loop_snapshot(value["resume_snapshot"])
    if (None if resume_snapshot is None else resume_snapshot["snapshot_digest"]) != control_snapshot_digest or (0 if resume_snapshot is None else resume_snapshot["generation"]) != control_generation:
        _fail("control-loop snapshot metadata is inconsistent")
    if value["ready_to_resume"] is not True:
        _fail("report is not ready to resume")
    if not isinstance(value["requires_external_reconciliation"], bool):
        _fail("external reconciliation marker is invalid")
    requires_reconciliation = any(row["from_phase"] == "dispatch_started" for row in recovered)
    if value["requires_external_reconciliation"] != requires_reconciliation:
        _fail("external reconciliation marker is inconsistent")
    body = {
        "schema": value["schema"],
        "status": value["status"],
        "active_count_before_recovery": active_count,
        "recovered": recovered,
        "recovery_digest": recovery_digest,
        "journal_snapshot_digest": journal_snapshot_digest,
        "journal_head_digest": journal_head_digest,
        "control_loop_snapshot_digest": control_snapshot_digest,
        "control_loop_generation": control_generation,
        "resume_snapshot": resume_snapshot,
        "ready_to_resume": value["ready_to_resume"],
        "requires_external_reconciliation": requires_reconciliation,
        "retention": value["retention"],
        "secret_material": value["secret_material"],
    }
    if require_digest and _digest(value.get("report_digest"), name="report_digest") != content_digest(body):
        _fail("report digest does not match its content")
    return {**body, "report_digest": content_digest(body)}


def validate_autonomous_goal_recovery_report(value: Mapping[str, Any]) -> dict[str, Any]:
    """Validate a sealed recovery report before it is stored or handed to a resume caller."""

    normalized = _report_body(value, require_digest=True)
    if len(canonical_json(normalized).encode("utf-8")) > MAX_GOAL_RECOVERY_REPORT_BYTES:
        _fail("report exceeds its bounded size")
    return copy.deepcopy(normalized)


class AutonomousGoalRecoveryCoordinator:
    """Compose journal recovery and loop resume with an explicit fail-closed ordering."""

    def __init__(
        self,
        ledger: AutonomousGoalLedger,
        journal: AutonomousGoalWorkerJournalPersistenceCoordinator,
        control_loop: AutonomousGoalControlLoopPersistenceCoordinator,
    ) -> None:
        if not isinstance(ledger, AutonomousGoalLedger):
            _fail("ledger must be an AutonomousGoalLedger")
        if not isinstance(journal, AutonomousGoalWorkerJournalPersistenceCoordinator):
            _fail("journal coordinator is invalid")
        if not isinstance(control_loop, AutonomousGoalControlLoopPersistenceCoordinator):
            _fail("control-loop coordinator is invalid")
        self.ledger = ledger
        self.journal = journal
        self.control_loop = control_loop
        self._report: dict[str, Any] | None = None

    @property
    def report(self) -> dict[str, Any] | None:
        return None if self._report is None else copy.deepcopy(self._report)

    def restore(self, *, now_ns: int | None = None) -> dict[str, Any]:
        if now_ns is not None:
            _integer(now_ns, name="now_ns")
        # This order is the safety boundary: reconcile and durably flush uncertain dispatches
        # before a stale control-loop snapshot can be used to select new work.
        journal_snapshot_before = self.journal.restore()
        active_before = self.journal.journal.active()
        recovered: list[dict[str, Any]] = []
        journal_snapshot = journal_snapshot_before
        if active_before:
            recovery = self.journal.journal.recover(self.ledger, now_ns=now_ns)
            raw_recovered = recovery.get("recovered")
            if not isinstance(raw_recovered, list):
                _fail("journal recovery returned malformed entries")
            recovered = [_entry(row, index=index) for index, row in enumerate(raw_recovered)]
            # A crash after this flush is safe: the next process sees only reconciled events.
            journal_snapshot = self.journal.flush()
        control_snapshot = self.control_loop.restore()
        status: RecoveryStatus = "recovered" if recovered else ("restored" if journal_snapshot_before is not None or control_snapshot is not None else "fresh")
        body = {
            "schema": GOAL_RECOVERY_SCHEMA,
            "status": status,
            "active_count_before_recovery": len(active_before),
            "recovered": recovered,
            "recovery_digest": content_digest(recovered),
            "journal_snapshot_digest": None if journal_snapshot is None else journal_snapshot["snapshot_digest"],
            "journal_head_digest": self.journal.journal.head_digest if journal_snapshot is None else journal_snapshot["head_digest"],
            "control_loop_snapshot_digest": None if control_snapshot is None else control_snapshot["snapshot_digest"],
            "control_loop_generation": 0 if control_snapshot is None else control_snapshot["generation"],
            "resume_snapshot": control_snapshot,
            "ready_to_resume": not bool(self.journal.journal.active()),
            "requires_external_reconciliation": any(row["from_phase"] == "dispatch_started" for row in recovered),
            "retention": GOAL_RECOVERY_RETENTION,
            "secret_material": "never_returned",
        }
        self._report = _report_body({**body, "report_digest": content_digest(body)}, require_digest=True)
        return self.report

    def assert_ready_for_resume(self) -> dict[str, Any]:
        if self._report is None:
            _fail("restore must complete before resume")
        if self.journal.journal.active():
            _fail("journal still contains active boundaries")
        return self.report

    def resume(self, loop: AutonomousGoalControlLoop, options: Mapping[str, Any] | None = None) -> AutonomousGoalControlLoopResult:
        if not isinstance(loop, AutonomousGoalControlLoop):
            _fail("resume requires an AutonomousGoalControlLoop")
        report = self.assert_ready_for_resume()
        safe_options = {} if options is None else dict(options)
        if "resume_snapshot" in safe_options:
            _fail("resume_snapshot is owned by the recovery coordinator")
        return loop.run(**safe_options, resume_snapshot=report["resume_snapshot"])

    def checkpoint(self, snapshot: Mapping[str, Any]) -> dict[str, Any]:
        """Persist one loop checkpoint with the worker journal durably ahead of it."""

        prior = self.assert_ready_for_resume()
        journal_snapshot = self.journal.flush()
        control_snapshot = self.control_loop.flush(snapshot)
        prior_body = {key: value for key, value in prior.items() if key != "report_digest"}
        body = {
            **prior_body,
            "journal_snapshot_digest": journal_snapshot["snapshot_digest"],
            "journal_head_digest": journal_snapshot["head_digest"],
            "control_loop_snapshot_digest": control_snapshot["snapshot_digest"],
            "control_loop_generation": control_snapshot["generation"],
            "resume_snapshot": control_snapshot,
        }
        self._report = _report_body({**body, "report_digest": content_digest(body)}, require_digest=True)
        return copy.deepcopy(control_snapshot)


__all__ = [
    "GOAL_RECOVERY_SCHEMA",
    "GOAL_RECOVERY_RETENTION",
    "MAX_GOAL_RECOVERY_GOALS",
    "MAX_GOAL_RECOVERY_REPORT_BYTES",
    "RecoveryStatus",
    "AutonomousGoalRecoveryCoordinator",
    "validate_autonomous_goal_recovery_report",
]

File: typescript/src/autonomous-goal-recovery.ts
import { ArgumentError, isObject } from "./errors.js";
import {
  AutonomousGoalControlLoop,
  type AutonomousGoalControlLoopResult,
} from "./autonomous-goal-control-loop.js";
import {
  AutonomousGoalControlLoopPersistenceCoordinator,
  validateAutonomousGoalControlLoopSnapshot,
  type AutonomousGoalControlLoopCheckpoint,
} from "./autonomous-goal-control-persistence.js";
import {
  AutonomousGoalWorkerJournalPersistenceCoordinator,
  type AutonomousGoalWorkerJournalPhase,
} from "./autonomous-goal-worker-journal.js";
import { InMemoryAutonomousGoalLedger } from "./autonomous-goals.js";
import { canonicalJson, digestJsonSync } from "./tooling.js";
import type { JsonObject } from "./types.js";

/** Ordered, metadata-only restart orchestration for long-horizon goal execution. */
export const AUTONOMOUS_GOAL_RECOVERY_SCHEMA = "bioprism-autonomous-goal-recovery/0.1" as const;
export const AUTONOMOUS_GOAL_RECOVERY_RETENTION = "metadata_only_goal_recovery;tasks_prompts_parameters_credentials_provider_values_and_results_not_retained" as const;
export const AUTONOMOUS_GOAL_RECOVERY_MAX_GOALS = 16_384;
export const AUTONOMOUS_GOAL_RECOVERY_MAX_REPORT_BYTES = 2_000_000;

export type AutonomousGoalRecoveryStatus = "fresh" | "restored" | "recovered";

export interface AutonomousGoalRecoveryEntry extends JsonObject {
  goal_id: string;
  from_phase: Extract<AutonomousGoalWorkerJournalPhase, "claimed" | "dispatch_started">;
  goal_status: "paused" | "blocked";
  outcome_digest: string;
}

export interface AutonomousGoalRecoveryReport extends JsonObject {
  schema: typeof AUTONOMOUS_GOAL_RECOVERY_SCHEMA;
  status: AutonomousGoalRecoveryStatus;
  active_count_before_recovery: number;
  recovered: AutonomousGoalRecoveryEntry[];
  recovery_digest: string;
  journal_snapshot_digest: string | null;
  journal_head_digest: string;
  control_loop_snapshot_digest: string | null;
  control_loop_generation: number;
  resume_snapshot: AutonomousGoalControlLoopCheckpoint | null;
  ready_to_resume: boolean;
  requires_external_reconciliation: boolean;
  retention: typeof AUTONOMOUS_GOAL_RECOVERY_RETENTION;
  secret_material: "never_returned";
  report_digest: string;
}

type GoalRecoveryResumeOptions = Omit<Parameters<AutonomousGoalControlLoop["run"]>[0], "resume_snapshot">;

function fail(message: string): never {
  throw new ArgumentError(`autonomous goal recovery ${message}`);
}

function digest(name: string, value: unknown, allowNull = false): string | null {
  if (value === null && allowNull) return null;
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) fail(`${name} must be a lowercase SHA-256 digest`);
  return value;
}

function integer(name: string, value: unknown, minimum = 0, maximum = Number.MAX_SAFE_INTEGER): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < minimum || value > maximum) fail(`${name} is outside its integer bounds`);
  return value;
}

function identifier(name: string, value: unknown, maximum = 256): string {
  if (typeof value !== "string" || !value.trim() || value.includes("\u0000") || new TextEncoder().encode(value).byteLength > maximum) fail(`${name} is outside its bounded identifier contract`);
  return value.trim();
}

function exactKeys(name: string, value: Record<string, unknown>, expected: readonly string[]): void {
  const allowed = new Set(expected);
  if (Object.keys(value).some((key) => !allowed.has(key)) || expected.some((key) => !(key in value))) fail(`${name} contains unsupported or missing fields`);
}

function normalizeEntry(value: unknown, index: number): AutonomousGoalRecoveryEntry {
  if (!isObject(value)) fail(`recovered entry ${index} is malformed`);
  exactKeys(`recovered entry ${index}`, value, ["goal_id", "from_phase", "goal_status", "outcome_digest"]);
  if (value.from_phase !== "claimed" && value.from_phase !== "dispatch_started") fail(`recovered entry ${index} phase is invalid`);
  if (value.goal_status !== "paused" && value.goal_status !== "blocked") fail(`recovered entry ${index} status is invalid`);
  return {
    goal_id: identifier(`recovered entry ${index} goal_id`, value.goal_id),
    from_phase: value.from_phase,
    goal_status: value.goal_status,
    outcome_digest: digest(`recovered entry ${index} outcome_digest`, value.outcome_digest)!,
  };
}

function reportBody(value: unknown, requireDigest: boolean): AutonomousGoalRecoveryReport {
  if (!isObject(value)) fail("report must be an object");
  const required = [
    "schema", "status", "active_count_before_recovery", "recovered", "recovery_digest", "journal_snapshot_digest",
    "journal_head_digest", "control_loop_snapshot_digest", "control_loop_generation", "resume_snapshot", "ready_to_resume",
    "requires_external_reconciliation", "retention", "secret_material",
  ];
  const allowed = new Set([...required, "report_digest"]);
  if (Object.keys(value).some((key) => !allowed.has(key)) || required.some((key) => !(key in value)) || (requireDigest && !("report_digest" in value))) fail("report contains unsupported or missing fields");
  if (value.schema !== AUTONOMOUS_GOAL_RECOVERY_SCHEMA || value.retention !== AUTONOMOUS_GOAL_RECOVERY_RETENTION || value.secret_material !== "never_returned") fail("report markers are invalid");
  if (value.status !== "fresh" && value.status !== "restored" && value.status !== "recovered") fail("report status is invalid");
  const activeCount = integer("active_count_before_recovery", value.active_count_before_recovery, 0, AUTONOMOUS_GOAL_RECOVERY_MAX_GOALS);
  if (!Array.isArray(value.recovered) || value.recovered.length > AUTONOMOUS_GOAL_RECOVERY_MAX_GOALS) fail("recovered entries are outside their bounds");
  const recovered = value.recovered.map(normalizeEntry);
  if (recovered.length !== activeCount) fail("recovered entries do not account for every active journal boundary");
  const goalIds = recovered.map((entry) => entry.goal_id);
  if (new Set(goalIds).size !== goalIds.length) fail("recovered entries contain duplicate goals");
  const recoveryDigest = digest("recovery_digest", value.recovery_digest)!;
  if (recoveryDigest !== digestJsonSync(recovered)) fail("recovery digest does not match recovered entries");
  const journalSnapshotDigest = digest("journal_snapshot_digest", value.journal_snapshot_digest, true);
  const journalHeadDigest = value.journal_head_digest === "" ? "" : digest("journal_head_digest", value.journal_head_digest)!;
  const controlSnapshotDigest = digest("control_loop_snapshot_digest", value.control_loop_snapshot_digest, true);
  const controlGeneration = integer("control_loop_generation", value.control_loop_generation, 0, 2_147_483_647);
  const resumeSnapshot = value.resume_snapshot === null ? null : validateAutonomousGoalControlLoopSnapshot(value.resume_snapshot as AutonomousGoalControlLoopCheckpoint);
  if ((resumeSnapshot?.snapshot_digest ?? null) !== controlSnapshotDigest || (resumeSnapshot?.generation ?? 0) !== controlGeneration) fail("control-loop snapshot metadata is inconsistent");
  if (typeof value.ready_to_resume !== "boolean" || value.ready_to_resume !== true) fail("report is not ready to resume");
  if (typeof value.requires_external_reconciliation !== "boolean") fail("external reconciliation marker is invalid");
  const requiresReconciliation = recovered.some((entry) => entry.from_phase === "dispatch_started");
  if (value.requires_external_reconciliation !== requiresReconciliation) fail("external reconciliation marker is inconsistent");
  const body = {
    schema: value.schema,
    status: value.status,
    active_count_before_recovery: activeCount,
    recovered,
    recovery_digest: recoveryDigest,
    journal_snapshot_digest: journalSnapshotDigest,
    journal_head_digest: journalHeadDigest,
    control_loop_snapshot_digest: controlSnapshotDigest,
    control_loop_generation: controlGeneration,
    resume_snapshot: resumeSnapshot,
    ready_to_resume: value.ready_to_resume,
    requires_external_reconciliation: requiresReconciliation,
    retention: value.retention,
    secret_material: value.secret_material,
  } satisfies Omit<AutonomousGoalRecoveryReport, "report_digest">;
  if (requireDigest && digest("report_digest", value.report_digest) !== digestJsonSync(body)) fail("report digest does not match its content");
  const normalized = {
    ...body,
    status: value.status as AutonomousGoalRecoveryStatus,
    retention: value.retention as typeof AUTONOMOUS_GOAL_RECOVERY_RETENTION,
    secret_material: "never_returned" as const,
  } satisfies Omit<AutonomousGoalRecoveryReport, "report_digest">;
  return { ...normalized, report_digest: digestJsonSync(normalized) };
}

export function validateAutonomousGoalRecoveryReport(value: unknown): AutonomousGoalRecoveryReport {
  const normalized = reportBody(value, true);
  if (new TextEncoder().encode(canonicalJson(normalized)).byteLength > AUTONOMOUS_GOAL_RECOVERY_MAX_REPORT_BYTES) fail("report exceeds its bounded size");
  return structuredClone(normalized);
}

/**
 * Restore the journal before the loop checkpoint, reconcile every uncertain boundary, and only
 * then expose a resume snapshot. This prevents a restarted loop from admitting work while an
 * earlier provider/effect invocation is still ambiguous. The caller still owns the ledger,
 * stores, task rehydrator, provider credentials, external reconciliation, and evaluator truth.
 */
export class AutonomousGoalRecoveryCoordinator {
  private reportValue: AutonomousGoalRecoveryReport | null = null;

  constructor(
    readonly ledger: InMemoryAutonomousGoalLedger,
    readonly journal: AutonomousGoalWorkerJournalPersistenceCoordinator,
    readonly control_loop: AutonomousGoalControlLoopPersistenceCoordinator,
  ) {
    if (!(ledger instanceof InMemoryAutonomousGoalLedger)) fail("ledger must be an InMemoryAutonomousGoalLedger");
    if (!(journal instanceof AutonomousGoalWorkerJournalPersistenceCoordinator)) fail("journal coordinator is invalid");
    if (!(control_loop instanceof AutonomousGoalControlLoopPersistenceCoordinator)) fail("control-loop coordinator is invalid");
  }

  get report(): AutonomousGoalRecoveryReport | null {
    return this.reportValue === null ? null : structuredClone(this.reportValue);
  }

  async restore(options: { now_ns?: number } = {}): Promise<AutonomousGoalRecoveryReport> {
    if (options.now_ns !== undefined) integer("now_ns", options.now_ns);
    // The order here is a correctness boundary. A control snapshot can be stale by one cycle,
    // but an active dispatch must never survive restart without a durable reconciliation event.
    const journalSnapshotBefore = await this.journal.restore();
    const activeBefore = this.journal.journal.active();
    let recovered: AutonomousGoalRecoveryEntry[] = [];
    let journalSnapshot = journalSnapshotBefore;
    if (activeBefore.length > 0) {
      const recovery = this.journal.journal.recover(this.ledger, { now_ns: options.now_ns });
      if (!Array.isArray(recovery.recovered)) fail("journal recovery returned malformed entries");
      recovered = recovery.recovered.map((entry, index) => normalizeEntry(entry, index));
      // Persist reconciliation before reading the control-loop checkpoint. A crash after this
      // write is safe: the next restore sees only terminal reconciled journal entries.
      journalSnapshot = await this.journal.flush();
    }
    const controlSnapshot = await this.control_loop.restore();
    const status: AutonomousGoalRecoveryStatus = recovered.length > 0 ? "recovered" : (journalSnapshotBefore !== null || controlSnapshot !== null ? "restored" : "fresh");
    const body = {
      schema: AUTONOMOUS_GOAL_RECOVERY_SCHEMA,
      status,
      active_count_before_recovery: activeBefore.length,
      recovered,
      recovery_digest: digestJsonSync(recovered),
      journal_snapshot_digest: journalSnapshot?.snapshot_digest ?? null,
      journal_head_digest: journalSnapshot?.head_digest ?? this.journal.journal.head_digest,
      control_loop_snapshot_digest: controlSnapshot?.snapshot_digest ?? null,
      control_loop_generation: controlSnapshot?.generation ?? 0,
      resume_snapshot: controlSnapshot,
      ready_to_resume: this.journal.journal.active().length === 0,
      requires_external_reconciliation: recovered.some((entry) => entry.from_phase === "dispatch_started"),
      retention: AUTONOMOUS_GOAL_RECOVERY_RETENTION,
      secret_material: "never_returned" as const,
    } satisfies Omit<AutonomousGoalRecoveryReport, "report_digest">;
    this.reportValue = reportBody({ ...body, report_digest: digestJsonSync(body) }, true);
    return structuredClone(this.reportValue);
  }

  assertReadyForResume(): AutonomousGoalRecoveryReport {
    if (this.reportValue === null) fail("restore must complete before resume");
    if (this.journal.journal.active().length > 0) fail("journal still contains active boundaries");
    return structuredClone(this.reportValue);
  }

  async resume(loop: AutonomousGoalControlLoop, options: GoalRecoveryResumeOptions = {}): Promise<AutonomousGoalControlLoopResult> {
    if (!(loop instanceof AutonomousGoalControlLoop)) fail("resume requires an AutonomousGoalControlLoop");
    const report = this.assertReadyForResume();
    const safeOptions = { ...options } as GoalRecoveryResumeOptions & { resume_snapshot?: unknown };
    if (Object.prototype.hasOwnProperty.call(safeOptions, "resume_snapshot")) fail("resume_snapshot is owned by the recovery coordinator");
    return loop.run({ ...safeOptions, resume_snapshot: report.resume_snapshot });
  }

  /**
   * Persist one completed loop checkpoint with the journal first. If control persistence fails,
   * the next restart still observes the settled journal and will not trust a stale loop cursor.
   */
  async checkpoint(snapshot: AutonomousGoalControlLoopCheckpoint): Promise<AutonomousGoalControlLoopCheckpoint> {
    const prior = this.assertReadyForResume();
    const journalSnapshot = await this.journal.flush();
    const controlSnapshot = await this.control_loop.flush(snapshot);
    const { report_digest: _reportDigest, ...priorBody } = prior;
    const body = {
      ...priorBody,
      journal_snapshot_digest: journalSnapshot.snapshot_digest,
      journal_head_digest: journalSnapshot.head_digest,
      control_loop_snapshot_digest: controlSnapshot.snapshot_digest,
      control_loop_generation: controlSnapshot.generation,
      resume_snapshot: controlSnapshot,
    } as Omit<AutonomousGoalRecoveryReport, "report_digest">;
    this.reportValue = reportBody({ ...body, report_digest: digestJsonSync(body) }, true);
    return structuredClone(controlSnapshot);
  }
}
